libs/browser/src/device.rs

```diff
@@ -0,0 +1,222 @@
+    name: "VxMark",
+    dimensions: (1080, 1920),
+};
+
+#[derive(Debug, Clone)]
+pub struct Selection {
+    values: Vec<Configuration>,
+    index: Arc<AtomicUsize>,
+}
+
+impl Selection {
+    pub fn new(selected_index: usize, values: impl IntoIterator<Item = Configuration>) -> Self {
+        let index = Arc::new(AtomicUsize::new(selected_index));
+        Self {
+            values: values.into_iter().collect(),
+            index,
+        }
+    }
+
+    pub fn current(&self) -> Configuration {
+        self.values[self.index.load(Ordering::Relaxed)]
+    }
+
+    pub fn current_index(&self) -> usize {
+        self.index.load(Ordering::Relaxed)
+    }
+
+    pub fn update_by_index(&self, index: usize) -> anyhow::Result<()> {
+        if index >= self.values.len() {
+            anyhow::bail!("Index out of bounds: {index}");
+        }
+        self.index.store(index, Ordering::Relaxed);
+        Ok(())
+    }
+
+    pub fn values(&self) -> &[Configuration] {
+        &self.values
+    }
+
+    pub fn update_by_value(&self, value: Configuration) -> anyhow::Result<()> {
+        let Some(index) = self.values.iter().position(|v| v == &value) else {
+            anyhow::bail!("Value not in list: {value:?}")
+        };
+
+        self.update_by_index(index)
+    }
+}
+
+impl Default for Selection {
+    fn default() -> Self {
+        Self::new(
+            0,
+            vec![
+                VXSCAN_CONFIG,
+                VXADMIN_CONFIG,
+                VXCENTRALSCAN_CONFIG,
+                VXMARK_CONFIG,
+            ],
+        )
+    }
+}
+
+/// Wrapper around a GTK window and a webview to ease various operations and
+/// state management.
+#[derive(Debug, Clone)]
+pub struct Device {
+    pub window: gtk::Window,
+    pub webview: WebView,
+    pub configuration: Arc<RwLock<Configuration>>,
+    pub is_inspector_open: Arc<AtomicBool>,
+}
+
+impl Device {
+    pub fn new(window: &gtk::Window, webview: &WebView, configuration: Configuration) -> Self {
+        let is_inspector_open = attach_inspector_handlers(&webview);
+
+        Self {
+            window: window.clone(),
+            webview: webview.clone(),
+            configuration: Arc::new(RwLock::new(configuration)),
+            is_inspector_open,
+        }
+    }
+
+    /// Returns the current configuration.
+    pub fn configuration(&self) -> anyhow::Result<Configuration> {
+        match self.configuration.try_read() {
+            Ok(reader) => Ok(*reader),
+            Err(e) => anyhow::bail!("Failed to read configuration: {e:?}"),
+        }
+    }
+
+    /// Resizes the window according to the current configuration.
+    pub fn reset_size(&self) -> anyhow::Result<()> {
+        let (width, height) = self.configuration()?.dimensions;
+        // reset minimum size
+        self.window.resize(1, 1);
+        // set the size we want for the webview; this will resize the window
+        self.webview.set_size_request(width, height);
```

Comment:
What happens if the monitor isn't big enough for the webview? Chrome's device emulation gives zoom options to deal with this

Reply:
I believe it just makes the window large enough that it doesn't fit within the display. I have it set up so that Parallels just draws the VM monitor at a high resolution but scaled down.

Reply:
Interesting, I guess I'll have to try it and see how it looks on my laptop screen